internal/graph/graph.go

```diff
@@ -0,0 +1,215 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package graph
+
+import (
+	"bytes"
+	"fmt"
+	"reflect"
+	"sync"
+
+	"github.com/pkg/errors"
+)
+
+var (
+	errArgKind = errors.New("constructor arguments must be pointers")
+
+	_typeOfError = reflect.TypeOf((*error)(nil)).Elem()
+)
+
+// Graph contains all Graph for current graph
+type Graph struct {
+	sync.RWMutex
+
+	nodes map[interface{}]graphNode
+}
+
+// NewGraph creates new data Graph for dig
+func NewGraph() Graph {
+	return Graph{
+		nodes: make(map[interface{}]graphNode),
+	}
+}
+
+// Reset the graph
+func (g *Graph) Reset() {
+	g.Lock()
+	defer g.Unlock()
+	g.nodes = make(map[interface{}]graphNode)
+}
+
+// Read reads value from the Graph
+func (g *Graph) Read(objType reflect.Type) (reflect.Value, error) {
+	// check if the type is a registered objNode
+	g.RLock()
+	n, ok := g.nodes[objType]
+	g.RUnlock()
+	if !ok {
+		return reflect.Zero(objType), fmt.Errorf("type %v is not registered", objType)
```

Comment:
There is no point of having a `reflect.Value` of specific type if there is an error :)
My comment was about unnecessary copying of reflect.Value that is going all around the graph. After digging through the code it looks like you'll need to change the graphNode interface to avoid it. May be another time then.